docs/correctness.md

```diff
@@ -0,0 +1,229 @@
+
+### Directory based operators
+
+Very core to our early design is that LLMs can trick the most clever evaluation infrastructure so relying purely on macro level evaluations is bound to miss things. So what we did was make it so all the kernels an LLM researcher need to generate kernels for are clear in a directory based structure.
+
+```bash
+generated_kernels
+├── add
+│   └── add_implementation_v1.py
+├── bitwise_and
+│   └── bitwise_and_implementation_v1.py
+├── div
+│   └── div_implementation_v1.py
+├── fmod
+│   └── fmod_implementation_v1.py
+├── mul
+│   └── mul_implementation_v1.py
+└── relu
+    └── relu_implementation_v1.py
+```
+
+That way if you do get promising speedups you can zip that folder and send it to your favorite performance engineer to debug. This protects us against a whole class of issues ranging from claiming victory when speedups exceed speed of light, bugs in the evaluation structure to some rogue agent making changes to the evaluation structure itself.
+
+
+## PyTorch benchmarking footguns
+
+Common benchmarking mistakes:
+
+1. **Async execution**: Basic timing only measures kernel launch, not completion
+   ```python
+   # Wrong: doesn't wait for GPU completion
+   tic = time.time()
+   model(inp)
+   toc = time.time()
+   ```
+
+2. **Warmup effects**: Running identical code sequentially gives misleading results due to caching
+   ```python
+   reference = eval(lambda a, b : a @ b, warmup=0)
+   llm_generation = eval(lambda a, b : a @ b, warmup=0)  # Faster due to cache!
+   ```
+
+**Solution**: Use `triton.testing.do_bench()` which handles kernel synchronization, warmups, and L2 cache clearing between runs.
+
+### Not all shapes are created equal
+
+Another common mistake is to pick small shapes when benchmarking along the lines of `model.forward(torch.randn(4,4,4,4))` the problem being is in this regime we're neither memory bandwidth bound or compute bound but instead overhead bound and the only thing we're really measuring here is the time it takes to launch a CUDA kernel in PyTorch. Note that small shapes are crucial for correctness we don't generally expect PyTorch to crash on weirdly shaped inputs.
+
+So the simplest fix to this problem is to pick the largest shapes we can, we can exponentially increase the sizes of our input tensors until we OOM and then slowly back off until we find a reasonably large shape. The results here will be more interesting than those with tiny shapes but fundamentally what most customers of PyTorch actually care about is performance on specific shapes i.e performance on important models.
+
+So we end up with a dataset that looks like this
+
+```
+Operator: aten.add.Tensor
+cnt: 156, ((T([1, 512, 768], f16), T([1, 512, 768], f16)), {})
+cnt: 89, ((T([32, 128], f32), T([32, 128], f32)), {})
```

Comment:
For the torchbench testsuite should I also add which models traces come from (this isn't too hard to do). Also it'd be wise to add a link to https://huggingface.co/datasets/GPUMODE/backendbench_tests